### PR TITLE
Optimise `Version` representation

### PR DESCRIPTION
**Pull request by hvr**
This optimises the [Int] representation to a 16-byte heap object for
~99% of version numbers occuring on Hackage.

One noteworthy improvement of this optimisation is a significant reduction of the size
of the `01-index.cache` file from previously 6299700 bytes (before #3905) down
to 3408864 bytes, i.e. down to ~54%

----

Also, this reduces the memory footprint a bit for e.g. `cabal info zzz` from

```
cabal.0: There is no package named 'zzz'.
You may need to run 'cabal update' to get the latest list of available
packages.
     859,337,368 bytes allocated in the heap
     447,261,128 bytes copied during GC
      37,385,208 bytes maximum residency (19 sample(s))
       1,311,136 bytes maximum slop
             103 MB total memory in use (0 MB lost due to fragmentation)

                                     Tot time (elapsed)  Avg pause  Max pause
  Gen  0      1613 colls,     0 par    0.268s   0.268s     0.0002s    0.0012s
  Gen  1        19 colls,     0 par    0.227s   0.227s     0.0119s    0.0506s

  TASKS: 4 (1 bound, 3 peak workers (3 total), using -N1)

  SPARKS: 0 (0 converted, 0 overflowed, 0 dud, 0 GC'd, 0 fizzled)

  INIT    time    0.001s  (  0.001s elapsed)
  MUT     time    0.431s  (  0.758s elapsed)
  GC      time    0.495s  (  0.495s elapsed)
  EXIT    time    0.006s  (  0.005s elapsed)
  Total   time    0.934s  (  1.259s elapsed)

  Alloc rate    1,991,870,623 bytes per MUT second

  Productivity  46.9% of total user, 34.8% of total elapsed
```

to

```
cabal.1: There is no package named 'zzz'.
You may need to run 'cabal update' to get the latest list of available
packages.
     834,314,392 bytes allocated in the heap
     440,791,176 bytes copied during GC
      36,663,112 bytes maximum residency (19 sample(s))
       2,225,040 bytes maximum slop
              96 MB total memory in use (0 MB lost due to fragmentation)

                                     Tot time (elapsed)  Avg pause  Max pause
  Gen  0      1574 colls,     0 par    0.254s   0.254s     0.0002s    0.0007s
  Gen  1        19 colls,     0 par    0.223s   0.223s     0.0118s    0.0474s

  TASKS: 4 (1 bound, 3 peak workers (3 total), using -N1)

  SPARKS: 0 (0 converted, 0 overflowed, 0 dud, 0 GC'd, 0 fizzled)

  INIT    time    0.001s  (  0.001s elapsed)
  MUT     time    0.383s  (  0.699s elapsed)
  GC      time    0.477s  (  0.477s elapsed)
  EXIT    time    0.005s  (  0.005s elapsed)
  Total   time    0.869s  (  1.182s elapsed)

  Alloc rate    2,175,866,164 bytes per MUT second

  Productivity  44.9% of total user, 33.0% of total elapsed
```

### CODE REVIEW

**Comment by dcoutts** [comment]
Should extend the tests in `Cabal/tests/UnitTests/Distribution/Version.hs`. Currently they cover `VersionRange` properties, but now that we have a non-trivial representation then I think we need proper coverage of `Version` itself.

So the things we need are: round trip tests, and Ord tests. These should have good coverage of versions with 0-6 components, and the boundary condition for the max size of each component.

**Comment by hvr** [comment]
@dcoutts  fwiw, here's some tests I used while developing:

```hs
prop_VersionId :: [Int] -> Bool
prop_VersionId lst0 = (versionNumbers . mkVersion) lst == lst
  where
    lst = map abs lst0

prop_VersionEq :: Base.Version -> Base.Version -> Bool
prop_VersionEq v1 v2 = (==) v1 v2 == ((==) `on` mkVersion') v1 v2

prop_VersionOrd :: Base.Version -> Base.Version -> Bool
prop_VersionOrd v1 v2 = compare v1 v2 == (compare `on` mkVersion') v1 v2
```

File: Cabal/Distribution/Version.hs [hunk 1]
```diff
@@ -138,7 +150,24 @@ instance Text Version where
...
-mkVersion = Version
+mkVersion ns = case ns of
+    [] -> nullVersion
+    [v1]          | v1 <= 0xfffe
+      -> PV0 (mkW64 (v1+1) 0 0 0)
+    [v1,v2]       | v1 <= 0xfffe, v2 <= 0xfffe
+      -> PV0 (mkW64 (v1+1) (v2+1) 0 0)
+    [v1,v2,v3]    | v1 <= 0xfffe, v2 <= 0xfffe, v3 <= 0xfffe
+      -> PV0 (mkW64 (v1+1) (v2+1) (v3+1)  0)
+    [v1,v2,v3,v4] | v1 <= 0xfffe, v2 <= 0xfffe, v3 <= 0xfffe, v4 <= 0xfffe
```
Comment (dcoutts) [inline Cabal/Distribution/Version.hs]: does it make any difference to do fewer comparisons, e.g. `v1 .|. v2 .|. v3 .|. v4 <= 0xfffe` (or is that trick correct since it's 0xfffe rather than 0xffff?)

File: Cabal/Distribution/Version.hs [hunk 2]
```diff
@@ -105,13 +107,23 @@ import Control.Exception (assert)
 -- 'Binary' instance using a different (and more compact) encoding.
 --
 -- @since 2.0
-data Version = Version [Int]
-             deriving (Data,Eq,Ord,Generic,Show,Read,Typeable)
+
+data Version = PV0 {-# UNPACK #-} !Word64
```
Comment (dcoutts) [inline Cabal/Distribution/Version.hs]: should add a non-haddock/internal comment to describe the representation trick

File: Cabal/Distribution/Version.hs (hunk 1 above)
Comment (hvr) [inline Cabal/Distribution/Version.hs]: I guess it would fail for e.g. v1=0x1, v2=0xfffe

**Comment by dcoutts** [comment]
@hvr those tests look sensible. Add them to the Version.hs test module. It may be necessary to use a newtype wrapper to use a different Arbitrary instance for [Int]. It should make sure we get coverage of the 0xffff boundary, and not get too many uselessly long lists.

**Comment by ezyang** [comment]
@hvr What do you need to do to make this non-WIP?

**Comment by hvr** [comment]
@ezyang I want to run a few criterion tests, look at runtime profiling reports, and maybe look at Core to see what kind of performance regression we may run into

**Comment by ezyang** [comment]
Sounds good. Obviously you don't have to, but it would be really awesome if this performance work could be recorded in some sort of durable way (ideally, running in the CI!) Allocations may be a decent, stable metric in this respect.

**Comment by hvr** [comment]
@ezyang @dcoutts fyi, I've updated the PR and removed the 'WIP' tag

File: Cabal/Distribution/Version.hs (hunk 2 above)
Comment (hvr) [inline Cabal/Distribution/Version.hs]: (still TODO)

File: Cabal/Distribution/Version.hs (hunk 1 above)
Comment (hvr) [inline Cabal/Distribution/Version.hs]: (still TODO)

File: Cabal/tests/UnitTests/Distribution/Version.hs [hunk 3]
```diff
@@ -147,6 +181,35 @@ instance Arbitrary VersionRange where
...
 
+---------------------
+-- Version properties
+--
+
+prop_VersionId :: [Int] -> Bool
+prop_VersionId lst0 =
+    (versionNumbers . mkVersion) lst == lst
+  where
+    lst = map abs lst0
```
Comment (dcoutts) [inline Cabal/tests/UnitTests/Distribution/Version.hs]: Can use `[NonNegative Int]` and `map getNonNegative` rather than `map abs` to be a bit clearer.

File: Cabal/Distribution/Version.hs (hunk 2 above)
Comment (dcoutts) [inline Cabal/Distribution/Version.hs]: In particular mention the invariant, since that explains why we can use a derived Eq (there's a canonical representation)

**Comment by dcoutts** [comment]
Great. Tests looks good.

File: Cabal/Distribution/Version.hs (hunk 1 above)
Comment (hvr) [inline Cabal/Distribution/Version.hs]: @dcoutts fwiw, I did benchmark, and it would result in slightly faster code... (and I had to take into account an additional check for `maxBound`s); however I'm saving this for a separate PR

File: Cabal/Distribution/Version.hs (hunk 2 above)
Comment (hvr) [inline Cabal/Distribution/Version.hs]: done

File: Cabal/Distribution/Version.hs (hunk 1 above)
Comment (hvr) [inline Cabal/Distribution/Version.hs]: @dcoutts please take a look at https://github.com/hvr/cabal-version-bench
